File: libs/python/stonecharts/charts/pie.py
"""Pie chart renderer: ChartSpec -> SVG string.

Centered sector arcs with value-to-angle mapping. Does NOT use
render_cartesian — pie has no axes (Family B polar foundation).
See charts/pie/design.md for the full geometry contract.
"""

from __future__ import annotations

import math

from ..spec import ChartSpec, Gradient
from ..util import esc, fmt_num
from ._cartesian import a11y_summary, gradient_def, pattern_def


def render_svg(spec: ChartSpec) -> str:
    W, H = spec.width, spec.height
    theme = spec.theme
    palette = theme.palette
    cid = esc(spec.id)

    a11y_attr = ""
    a11y_desc = ""
    if spec.a11y:
        _sum = esc(a11y_summary(spec, "Pie"))
        a11y_attr = f' role="img" aria-label="{_sum}"'
        a11y_desc = f"<desc>{_sum}</desc>"

    m_top: float = 20
    if spec.title:
        m_top += 26
    if spec.subtitle:
        m_top += 18
    m_left: float = 22
    m_right: float = 22
    m_bottom: float = 28 + (18 if spec.legend else 0)
    if spec.layout and spec.layout.margin:
        m = spec.layout.margin
        if m.top is not None:
            m_top = m.top
        if m.left is not None:
            m_left = m.left
        if m.right is not None:
            m_right = m.right
        if m.bottom is not None:
            m_bottom = m.bottom

    plot_x, plot_y = m_left, m_top
    plot_w, plot_h = W - m_left - m_right, H - m_top - m_bottom

    s0 = spec.series[0] if spec.series else None
    data = list(s0.data) if s0 else []
    n = len(data)
    cats = spec.x_axis.categories or [str(i) for i in range(n)]

    defs_parts: list[str] = []
    color_by_point = True
    pie_fill = ""
    if s0 is not None:
        if isinstance(s0.color, Gradient):
            gid = f"{cid}-grad-0"
            defs_parts.append(gradient_def(gid, s0.color))
            pie_fill = f"url(#{gid})"
            color_by_point = False
        elif s0.color:
            pie_fill = esc(s0.color)
            color_by_point = False
        if s0.pattern is not None:
            pid = f"{cid}-pat-0"
            defs_parts.append(pattern_def(pid, s0.pattern))
            pie_fill = f"url(#{pid})"
            color_by_point = False

    solid0 = ""
    if s0 is not None:
        if isinstance(s0.color, Gradient):
            solid0 = esc(s0.color.stops[0].color) if s0.color.stops else esc(palette[0])
        elif s0.color:
            solid0 = esc(s0.color)
        else:
            solid0 = esc(palette[0])

    total = sum(v for v in data if v > 0)

    stroke_color = "#1e1e2f" if theme.background else "#ffffff"

    p: list[str] = []

    _font = 'font-family="Segoe UI, Helvetica, Arial, sans-serif"'
    if spec.responsive:
        p.append(
            f'<svg class="sc-chart"{a11y_attr} xmlns="http://www.w3.org/2000/svg" '
            f'viewBox="0 0 {W} {H}" preserveAspectRatio="xMidYMid meet" '
            f'width="100%" {_font}>'
        )
    else:
        p.append(
            f'<svg class="sc-chart"{a11y_attr} xmlns="http://www.w3.org/2000/svg" '
            f'width="{W}" height="{H}" viewBox="0 0 {W} {H}" {_font}>'
        )

    if a11y_desc:
        p.append(a11y_desc)

    if defs_parts:
        p.append("<defs>" + "".join(defs_parts) + "</defs>")

    if theme.background:
        p.append(f'<rect class="sc-bg" x="0" y="0" width="{W}" height="{H}" fill="{theme.background}"/>')

    ty = 26
    if spec.title:
        p.append(
            f'<text class="sc-title" x="{W / 2:.1f}" y="{ty}" text-anchor="middle" '
            f'font-size="17" font-weight="600" fill="{theme.title_color}">{esc(spec.title)}</text>'
        )
        ty += 20
    if spec.subtitle:
        p.append(
            f'<text class="sc-subtitle" x="{W / 2:.1f}" y="{ty}" text-anchor="middle" '
            f'font-size="12" fill="{theme.subtitle_color}">{esc(spec.subtitle)}</text>'
        )

    p.append(
        f'<line class="sc-crosshair" x1="0" y1="{plot_y:.1f}" x2="0" y2="{plot_y + plot_h:.1f}" '
        f'stroke="{theme.crosshair_color}" stroke-width="1" stroke-dasharray="4 3" style="display:none"/>'
    )

    cx = plot_x + plot_w / 2
    cy = plot_y + plot_h / 2
    r = min(plot_w, plot_h) / 2

    if n > 0 and s0 is not None and total > 0:
        p.append('<g class="sc-series" data-series="0">')

        positive_count = sum(1 for v in data if v > 0)

        if positive_count == 1:
            idx = next(i for i, v in enumerate(data) if v > 0)
            fill = esc(palette[idx % len(palette)]) if color_by_point else pie_fill
            cat = cats[idx] if idx < len(cats) else str(idx)
            pct = f"{100.0:.1f}%"
            p.append(
                f'<circle class="sc-slice sc-point" data-series="0" '
                f'data-series-name="{esc(s0.name)}" data-x="{esc(cat)}" '
                f'data-y="{esc(fmt_num(data[idx]))}" data-color="{fill}" '
                f'data-index="{idx}" data-percentage="{pct}" '
                f'data-r="{r:.1f}" data-r-hover="{r + 4:.1f}" '
                f'cx="{cx:.1f}" cy="{cy:.1f}" r="{r:.1f}" '
                f'fill="{fill}" stroke="{stroke_color}" stroke-width="2"/>'
            )
        else:
            angle = -math.pi / 2
            for i in range(n):
                v = data[i]
                if v <= 0:
                    continue
                sweep = (v / total) * 2 * math.pi
                x1 = cx + r * math.cos(angle)
                y1 = cy + r * math.sin(angle)
                end_angle = angle + sweep
                x2 = cx + r * math.cos(end_angle)
                y2 = cy + r * math.sin(end_angle)
                large_arc = 1 if sweep > math.pi else 0

                fill = esc(palette[i % len(palette)]) if color_by_point else pie_fill
                cat = cats[i] if i < len(cats) else str(i)
                pct = f"{(v / total) * 100:.1f}%"

                p.append(
                    f'<path class="sc-slice sc-point" data-series="0" '
                    f'data-series-name="{esc(s0.name)}" data-x="{esc(cat)}" '
                    f'data-y="{esc(fmt_num(v))}" data-color="{fill}" '
                    f'data-index="{i}" data-percentage="{pct}" '
                    f'data-r="{r:.1f}" data-r-hover="{r + 4:.1f}" '
                    f'cx="{cx:.1f}" cy="{cy:.1f}" '
                    f'd="M {cx:.1f} {cy:.1f} L {x1:.1f} {y1:.1f} '
                    f'A {r:.1f} {r:.1f} 0 {large_arc} 1 {x2:.1f} {y2:.1f} Z" '
                    f'fill="{fill}" stroke="{stroke_color}" stroke-width="2"/>'
                )

                angle = end_angle

        p.append("</g>")

    if spec.legend and spec.series and s0 is not None:
        gap = 22
        est = [len(s0.name) * 7 + 26]
        total_w = sum(est) + gap * (len(est) - 1) if est else 0
        lx = plot_x + (plot_w - total_w) / 2
        ly = H - 10
        p.append('<g class="sc-legend">')
        p.append('<g class="sc-legend-item" data-series="0">')
        p.append(f'<rect x="{lx:.1f}" y="{ly - 9:.1f}" width="14" height="4" rx="2" fill="{solid0}"/>')
        p.append(
            f'<text x="{lx + 20:.1f}" y="{ly - 2:.1f}" font-size="12" '
            f'fill="{theme.legend_text_color}">{esc(s0.name)}</text>'
        )
        p.append("</g>")
        p.append("</g>")

    p.append("</svg>")
    return "".join(p)

File: libs/go/pie.go
package stonecharts

import (
	"fmt"
	"math"
	"strings"
	"unicode/utf8"
)

func renderPieSVG(spec *ChartSpec) string {
	W, H := int(spec.Width), int(spec.Height)
	theme := spec.theme
	if theme == nil {
		t := lightTheme()
		theme = &t
	}
	palette := theme.Palette
	cid := esc(spec.ID)

	a11yAttr := ""
	a11yDesc := ""
	if spec.a11yOn() {
		sum := esc(a11ySummary(spec, "Pie"))
		a11yAttr = fmt.Sprintf(` role="img" aria-label="%s"`, sum)
		a11yDesc = fmt.Sprintf("<desc>%s</desc>", sum)
	}

	mTop := 20.0
	if spec.Title != "" {
		mTop += 26
	}
	if spec.Subtitle != "" {
		mTop += 18
	}
	mLeft := 22.0
	mRight := 22.0
	mBottom := 28.0
	if spec.legendOn() {
		mBottom += 18
	}
	if spec.Layout != nil && spec.Layout.Margin != nil {
		m := spec.Layout.Margin
		if m.Top != nil {
			mTop = *m.Top
		}
		if m.Left != nil {
			mLeft = *m.Left
		}
		if m.Right != nil {
			mRight = *m.Right
		}
		if m.Bottom != nil {
			mBottom = *m.Bottom
		}
	}

	plotX, plotY := mLeft, mTop
	plotW, plotH := float64(W)-mLeft-mRight, float64(H)-mTop-mBottom

	var s0 *Series
	if len(spec.Series) > 0 {
		s0 = &spec.Series[0]
	}
	var data []float64
	if s0 != nil {
		data = s0.Data
	}
	n := len(data)
	cats := spec.XAxis.Categories
	if len(cats) == 0 {
		cats = make([]string, n)
		for i := 0; i < n; i++ {
			cats[i] = fmt.Sprintf("%d", i)
		}
	}

	var defsParts []string
	colorByPoint := true
	pieFill := ""
	if s0 != nil {
		grad, solidHex := s0.colorSpec()
		if grad != nil {
			gid := cid + "-grad-0"
			defsParts = append(defsParts, gradientDef(gid, grad))
			pieFill = "url(#" + gid + ")"
			colorByPoint = false
		} else if solidHex != "" {
			pieFill = esc(solidHex)
			colorByPoint = false
		}
		if s0.Pattern != nil {
			pid := cid + "-pat-0"
			defsParts = append(defsParts, patternDef(pid, s0.Pattern))
			pieFill = "url(#" + pid + ")"
			colorByPoint = false
		}
	}

	solid0 := ""
	if s0 != nil {
		grad, solidHex := s0.colorSpec()
		if grad != nil {
			if len(grad.Stops) > 0 {
				solid0 = esc(grad.Stops[0].Color)
			} else {
				solid0 = esc(palette[0])
			}
		} else if solidHex != "" {
			solid0 = esc(solidHex)
		} else {
			solid0 = esc(palette[0])
		}
	}

	total := 0.0
	for _, v := range data {
		if v > 0 {
			total += v
		}
	}

	strokeColor := "#ffffff"
	if theme.Background != "" {
		strokeColor = "#1e1e2f"
	}

	var p strings.Builder

	font := `font-family="Segoe UI, Helvetica, Arial, sans-serif"`
	if spec.Responsive {
		p.WriteString(fmt.Sprintf(
			`<svg class="sc-chart"%s xmlns="http://www.w3.org/2000/svg" viewBox="0 0 %d %d" preserveAspectRatio="xMidYMid meet" width="100%%" %s>`,
			a11yAttr, W, H, font))
	} else {
		p.WriteString(fmt.Sprintf(
			`<svg class="sc-chart"%s xmlns="http://www.w3.org/2000/svg" width="%d" height="%d" viewBox="0 0 %d %d" %s>`,
			a11yAttr, W, H, W, H, font))
	}

	if a11yDesc != "" {
		p.WriteString(a11yDesc)
	}

	if len(defsParts) > 0 {
		p.WriteString("<defs>" + strings.Join(defsParts, "") + "</defs>")
	}

	if theme.Background != "" {
		p.WriteString(fmt.Sprintf(
			`<rect class="sc-bg" x="0" y="0" width="%d" height="%d" fill="%s"/>`,
			W, H, theme.Background))
	}

	ty := 26
	if spec.Title != "" {
		p.WriteString(fmt.Sprintf(
			`<text class="sc-title" x="%s" y="%d" text-anchor="middle" font-size="17" font-weight="600" fill="%s">%s</text>`,
			f1(float64(W)/2), ty, theme.TitleColor, esc(spec.Title)))
		ty += 20
	}
	if spec.Subtitle != "" {
		p.WriteString(fmt.Sprintf(
			`<text class="sc-subtitle" x="%s" y="%d" text-anchor="middle" font-size="12" fill="%s">%s</text>`,
			f1(float64(W)/2), ty, theme.SubtitleColor, esc(spec.Subtitle)))
	}

	p.WriteString(fmt.Sprintf(
		`<line class="sc-crosshair" x1="0" y1="%s" x2="0" y2="%s" stroke="%s" stroke-width="1" stroke-dasharray="4 3" style="display:none"/>`,
		f1(plotY), f1(plotY+plotH), theme.CrosshairColor))

	cx := plotX + plotW/2
	cy := plotY + plotH/2
	r := plotW / 2
	if plotH/2 < r {
		r = plotH / 2
	}

	if n > 0 && s0 != nil && total > 0 {
		p.WriteString(`<g class="sc-series" data-series="0">`)

		positiveCount := 0
		for _, v := range data {
			if v > 0 {
				positiveCount++
			}
		}

		if positiveCount == 1 {
			idx := 0
			for i, v := range data {
				if v > 0 {
					idx = i
					break
				}
			}
			var fill string
			if colorByPoint {
				fill = esc(palette[idx%len(palette)])
			} else {
				fill = pieFill
			}
			cat := fmt.Sprintf("%d", idx)
			if idx < len(cats) {
				cat = cats[idx]
			}
			pct := "100.0%"
			p.WriteString(fmt.Sprintf(
				`<circle class="sc-slice sc-point" data-series="0" data-series-name="%s" data-x="%s" data-y="%s" data-color="%s" data-index="%d" data-percentage="%s" data-r="%s" data-r-hover="%s" cx="%s" cy="%s" r="%s" fill="%s" stroke="%s" stroke-width="2"/>`,
				esc(s0.Name), esc(cat), esc(fmtNum(data[idx])), fill,
				idx, pct, f1(r), f1(r+4),
				f1(cx), f1(cy), f1(r),
				fill, strokeColor))
		} else {
			angle := -math.Pi / 2
			for i := 0; i < n; i++ {
				v := data[i]
				if v <= 0 {
					continue
				}
				sweep := (v / total) * 2 * math.Pi
				x1 := cx + r*math.Cos(angle)
				y1 := cy + r*math.Sin(angle)
				endAngle := angle + sweep
				x2 := cx + r*math.Cos(endAngle)
				y2 := cy + r*math.Sin(endAngle)
				largeArc := 0
				if sweep > math.Pi {
					largeArc = 1
				}

				var fill string
				if colorByPoint {
					fill = esc(palette[i%len(palette)])
				} else {
					fill = pieFill
				}
				cat := fmt.Sprintf("%d", i)
				if i < len(cats) {
					cat = cats[i]
				}
				pct := fmt.Sprintf("%.1f%%", (v/total)*100)

				p.WriteString(fmt.Sprintf(
					`<path class="sc-slice sc-point" data-series="0" data-series-name="%s" data-x="%s" data-y="%s" data-color="%s" data-index="%d" data-percentage="%s" data-r="%s" data-r-hover="%s" cx="%s" cy="%s" d="M %s %s L %s %s A %s %s 0 %d 1 %s %s Z" fill="%s" stroke="%s" stroke-width="2"/>`,
					esc(s0.Name), esc(cat), esc(fmtNum(v)), fill,
					i, pct, f1(r), f1(r+4),
					f1(cx), f1(cy),
					f1(cx), f1(cy), f1(x1), f1(y1),
					f1(r), f1(r), largeArc, f1(x2), f1(y2),
					fill, strokeColor))

				angle = endAngle
			}
		}

		p.WriteString("</g>")
	}

	if spec.legendOn() && len(spec.Series) > 0 {
		gap := 22.0
		var est []float64
		if s0 != nil {
			est = []float64{float64(utf8.RuneCountInString(s0.Name)*7 + 26)}
		}
		totalW := 0.0
		if len(est) > 0 {
			for _, e := range est {
				totalW += e
			}
			totalW += gap * float64(len(est)-1)
		}
		lx := plotX + (plotW-totalW)/2
		ly := float64(H - 10)
		p.WriteString(`<g class="sc-legend">`)
		p.WriteString(`<g class="sc-legend-item" data-series="0">`)
		p.WriteString(fmt.Sprintf(
			`<rect x="%s" y="%s" width="14" height="4" rx="2" fill="%s"/>`,
			f1(lx), f1(ly-9), solid0))
		p.WriteString(fmt.Sprintf(
			`<text x="%s" y="%s" font-size="12" fill="%s">%s</text>`,
			f1(lx+20), f1(ly-2), theme.LegendTextColor, esc(s0.Name)))
		p.WriteString("</g>")
		p.WriteString("</g>")
	}

	p.WriteString("</svg>")
	return p.String()
}

func pieDataTable(spec *ChartSpec) string {
	var b strings.Builder
	b.WriteString(`<table class="sc-visually-hidden">`)
	if spec.Title != "" {
		b.WriteString("<caption>" + esc(spec.Title) + "</caption>")
	}
	b.WriteString(`<thead><tr><th scope="col">Category</th><th scope="col">Value</th>` +
		`<th scope="col">Percentage</th></tr></thead><tbody>`)
	cats := spec.XAxis.Categories
	if len(spec.Series) > 0 {
		s0 := spec.Series[0]
		total := 0.0
		for _, v := range s0.Data {
			if v > 0 {
				total += v
			}
		}
		for i, v := range s0.Data {
			cat := fmt.Sprintf("%d", i)
			if i < len(cats) {
				cat = cats[i]
			}
			pct := 0.0
			if total > 0 {
				pct = (v / total) * 100
			}
			b.WriteString(fmt.Sprintf(
				`<tr><th scope="row">%s</th><td>%s</td><td>%.1f%%</td></tr>`,
				esc(cat), esc(fmtNum(v)), pct))
		}
	}
	b.WriteString("</tbody></table>")
	return b.String()
}
